### PR TITLE
add register useless response to multiple tasks

### PR DESCRIPTION
**Pull request by pinges**
## PR description
Some of the tasks in Besu, especially the Snapsync ones, did not register a useless response when necessary. When a specific number of useless reponses have been registered against an EthPeer within a certain time, the EthPeer is disconnected.
Without this mechanism it is possible that Besu is looping through useless peers for a long time.
With the change, useless peers will be disconnected, which makes room for Besu to find new, hopefully useful, peers.

### CODE REVIEW

**Comment by matkt** [comment]
can try to simulate a peer that returns an empty réponse each time on Goerli and see if the node can be unblocked. The best peer response should be changed by ensuring that it response is transformed into an empty as soon as the snapsync has already downloaded a few datas. To be sure that it is not a check before the Snapsync which unlocks

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetStorageRangeFromPeerTask.java [hunk 1]
```diff
@@ -75,6 +76,12 @@ protected CompletableFuture<StorageRangeMessage.SlotRangeData> executePeerTask(
     return executeSubTask(task::run)
         .thenApply(
             peerResult -> {
+              if (isEmptyResponse(peerResult.getResult())) {
+                final EthPeer peer = peerResult.getPeer();
+                peer.recordUselessResponse("GetAccountRangeFromPeerTask");
```
[the remark this answers is not in the document]
Comment (macfarla) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetStorageRangeFromPeerTask.java:81]: can this be abstracted out easily? it's almost duplicated

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java [hunk 2]
```diff
@@ -147,6 +147,12 @@ protected boolean isPeerFailure(final Throwable error) {
     return error instanceof PeerBreachedProtocolException
         || error instanceof PeerDisconnectedException
         || error instanceof NoAvailablePeersException;
+    // should we add this here?:
+    //        || error instanceof IncompleteResultsException;
+  }
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java:152]: ? what are the implications of this

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/ethtaskutils/AbstractMessageTaskTestBase.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright ConsenSys AG.
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/ethtaskutils/AbstractMessageTaskTestBase.java:2]:
```suggestion
 * Copyright contributors to Hyperledger Besu.
```

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTaskTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    // Execute task and wait for response
+    final EthTask<AccountRangeMessage.AccountRangeData> task = createTask();
+    final CompletableFuture<AccountRangeMessage.AccountRangeData> future = task.run();
+
+    assertThat(respondingEthPeer.getEthPeer().getReputation().getScore()).isEqualTo(100);
+    // peer responds with empty response
+    respondingEthPeer.respond(RespondingEthPeer.emptyResponderForSnap());
+
+    assertThat(future.isDone()).isTrue();
+    assertThat(respondingEthPeer.getEthPeer().getReputation().getScore()).isEqualTo(99);
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTaskTest.java:61]: yay tests

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java [hunk 5]
```diff
@@ -97,9 +113,100 @@ public void getHeadersFromHashWithSkipReversed() {
...
+    final EthTask<AbstractPeerTask.PeerTaskResult<List<BlockHeader>>> task =
+        createTask(requestedData);
+    final CompletableFuture<AbstractPeerTask.PeerTaskResult<List<BlockHeader>>> future = task.run();
+    respondingEthPeer.respond(responder);
+    assertThat(future.isDone()).isTrue();
+    assertThat(future.isCompletedExceptionally()).isFalse();
+    assertThat(respondingEthPeer.getEthPeer().getReputation().getScore()).isEqualTo(99);
+  }
+
+  private List<BlockHeader> dummyBHList(final int count) {
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java:136]: nit: I think generally we have helper methods at the bottom, after all the `@Tests`

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java [hunk 6]
```diff
@@ -97,9 +113,100 @@ public void getHeadersFromHashWithSkipReversed() {
...
+    final CompletableFuture<AbstractPeerTask.PeerTaskResult<List<BlockHeader>>> future = task.run();
+    respondingEthPeer.respond(
+        (c, m) -> Optional.of(BlockHeadersMessage.create(dummyBHList(DEFAULT_COUNT + 1))));
+    assertThat(future.isDone()).isTrue();
+    assertThat(future.isCompletedExceptionally()).isFalse();
+    assertThat(respondingEthPeer.getEthPeer().getReputation().getScore()).isEqualTo(99);
+  }
+
+  @Test
+  public void completesWhenPeersSendsWrongFirstHeaderAndReducesReputation() {
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java:187]:
```suggestion
  public void completesWhenPeerSendsWrongFirstHeaderAndReducesReputation() {
```

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java [hunk 7]
```diff
@@ -97,9 +113,100 @@ public void getHeadersFromHashWithSkipReversed() {
...
+              Hash.ZERO,
+              DataGas.fromHexString("0x1234"),
+              new MainnetBlockHeaderFunctions(),
+              Optional.empty()));
+    }
+    return blockHeaders;
+  }
+
+  @Test
+  public void completesWhenPeersSendsTooManyHeadersAndReducesReputation() {
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java:166]:
```suggestion
  public void completesWhenPeerSendsTooManyHeadersAndReducesReputation() {
```

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java [hunk 8]
```diff
@@ -195,9 +302,87 @@ public void checkThatSequentialHeadersNotFormingAChainFails() {
...
   }
+
+  @Test
+  public void checkThatStreamClosedResponseCallsPeerRecordUselessResponse() {
+    final EthPeer peer = mock(EthPeer.class);
+    final AbstractGetHeadersFromPeerTask task =
+        new GetHeadersFromPeerByHashTask(
+            protocolSchedule, ethContext, Hash.ZERO, 0, 2, 0, false, metricsSystem);
+    task.processResponse(true, BlockHeadersMessage.create(Collections.emptyList()), peer);
+    Mockito.verify(peer, Mockito.times(1)).recordUselessResponse("headers");
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java:318]: this message seems a little brief

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java [hunk 9]
```diff
@@ -71,7 +72,14 @@ protected CompletableFuture<AccountRangeMessage.AccountRangeData> executePeerTas
...
         .thenApply(
             peerResult -> {
+              if (isEmptyResponse(peerResult.getResult())) {
+                final EthPeer peer = peerResult.getPeer();
+                peer.recordUselessResponse("GetAccountRangeFromPeerTask");
+                throw new IncompleteResultsException(
+                    "No account and proof returned by peer " + peer.nodeId());
+              }
               result.complete(peerResult.getResult());
+              // TODO: update block height of peer?
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java:82]: I hope that I will get some feedback about that from Fabio, when we are talking tonight.

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTaskTest.java (hunk 4 above)
Comment (pinges) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTaskTest.java:61]: :-)

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java (hunk 2 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java:152]: This is something I need to talk to Karim about!

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java (hunk 5 above)
Comment (pinges) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java:136]: done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java]: I've added the update of the estimated hight to the tasks

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java (hunk 2 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java:150]: Removed the comment.

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java (hunk 8 above)
Comment (pinges) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java:289]: Fixed

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetTrieNodeFromPeerTask.java [hunk 10]
```diff
@@ -64,7 +65,14 @@ protected CompletableFuture<Map<Bytes, Bytes>> executePeerTask(
     return executeSubTask(task::run)
         .thenApply(
             peerResult -> {
+              if (isEmptyResponse(peerResult.getResult())) {
+                final EthPeer peer = peerResult.getPeer();
+                peer.recordUselessResponse("GetTrieNodeFromPeerTask");
+                throw new IncompleteResultsException(
```
Comment (matkt) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetTrieNodeFromPeerTask.java:71]: maybe more result.completeExceptionally() instead of throwing direclty

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java (hunk 2 above)
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java:150]: I think we should do as here 
https://github.com/hyperledger/besu/blob/main/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/tasks/RetryingGetHeaderFromPeerByHashTask.java#L106 it depends if we want to use AbstractRetryingSwitchingPeerTask or not because retrying with the same peers will not make sens

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java [hunk 11]
```diff
@@ -71,7 +72,14 @@ protected CompletableFuture<AccountRangeMessage.AccountRangeData> executePeerTas
...
         .thenApply(
             peerResult -> {
+              if (isEmptyResponse(peerResult.getResult())) {
+                final EthPeer peer = peerResult.getPeer();
+                peer.recordUselessResponse("GetAccountRangeFromPeerTask");
+                throw new IncompleteResultsException(
+                    "No account range returned by peer " + peer.nodeId());
+              }
               result.complete(peerResult.getResult());
+              peerResult.getPeer().chainState().updateHeightEstimate(blockHeader.getNumber());
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java:82]: The estimate is only updated if the new height is greater than the current one.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractGetHeadersFromPeerTask.java
Comment (jflo) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractGetHeadersFromPeerTask.java:1]: Big improvements!

**Review by jflo — APPROVED** [review]

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java:82]: ok seems to be a good idea to fix some of the pivot issues we have. @fab-10  had also some ideas regarding how to fix the bad height issues

File: ethereum/evmtool/txs.rlp [hunk 12]
```diff
@@ -0,0 +1 @@
...e122131cdb47d6c0c77b371f8600a0a8307a12094000000000000000000000000000000000000010a80801ba0a3a2018ab0bc2695b94bb85d710f4d07132a94f8c3e0f385824da5fee11899a5a00d2dfe430ea5aaff3de8bbb9339e7485474c8e4e34636f787124a7a91e4d6d6af8600b0a8307a12094000000000000000000000000000000000000010b80801ba0b91968fdb3aecea26094ec30649daa4de81a875bcb1a123e732b8f3f112ce232a02ef8cd85969d8bcef5f4ee1f5d20783b8d9b7466726c15ebf911565825187665f8600c0a8307a12094000000000000000000000000000000000000010c80801ca0dd27e75aa990793205805c22265b04be8299b208fad4f37a7f652ecf32b67390a05aa8cda18521548ff8f95e88f49f309d05cab32de28a0942b8a7a824c50df459f8600d0a8307a12094000000000000000000000000000000000000010d80801ba0fad07ce7139dd4e00266194e6a51c048f74eaba3c0a1b03ece378a810abfaa63a04fec880dafaa5382797b4f88b16138b1f0c4e084817072c77ff9bf17ddd4ac26f8600e0a8307a12094000000000000000000000000000000000000010e80801ca0208b22ab245221bdc5cae6586d2ef019a2c37be41166e04b8abe354c41a8f5b6a032d5d6ef07731cd1684531c775c1727ef6ba75de18cda96d998aaa0c1db0bd68f8600f0a8307a12094000000000000000000000000000000000000010f80801ba0055225ffd3d8b2d19c32aa68cb46e7b52c1d99844fb8b7a53b922ea1649e9c5ba06ae2a1e3b9712354b706d0f4da6ea76ed2f8f75277a51a14a3e0ccf25b85c626"
```
Comment (matkt) [inline ethereum/evmtool/txs.rlp:1]: could you help me to understand why we need to add that for this PR ?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/BlockPropagationManager.java [hunk 13]
```diff
@@ -496,7 +496,8 @@ private CompletableFuture<Block> maybeRepeatGetBlock(
             .orElseGet(() -> blockchain.getBlockByNumber(blockNumber));
 
     // check if we got this block by other means
-    if (maybeBlock.isPresent()) {
+    if (maybeBlock.isPresent()
+        || (maybeBlockHash.isPresent() && pendingBlocksManager.contains(maybeBlockHash.get()))) {
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/BlockPropagationManager.java:500]: this seems unrelated to peering - is there tests around this?

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java [hunk 14]
```diff
@@ -97,9 +108,70 @@ public void getHeadersFromHashWithSkipReversed() {
...
+    assertThat(future.isDone()).isTrue();
+    assertThat(future.isCompletedExceptionally()).isFalse();
+    assertThat(respondingEthPeer.getEthPeer().getReputation().getScore()).isEqualTo(99);
+  }
+
+  @Test
+  public void completesWhenPeerSendsWrongFirstHeaderAndReducesReputation() {
+    // Setup a peer returning result where the first header has the wrong number
+    final RespondingEthPeer respondingEthPeer =
+        EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 1000);
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java:155]: what makes the first header have the wrong number?

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java [hunk 15]
```diff
@@ -97,9 +108,70 @@ public void getHeadersFromHashWithSkipReversed() {
...
+    // Setup data to be requested
+    final List<BlockHeader> requestedData =
+        generateDataToBeRequested(); // request DEFAULT_COUNT headers
+
+    // Execute task and wait for response
+    final EthTask<AbstractPeerTask.PeerTaskResult<List<BlockHeader>>> task =
+        createTask(requestedData);
+    final CompletableFuture<AbstractPeerTask.PeerTaskResult<List<BlockHeader>>> future = task.run();
+    respondingEthPeer.respond(
+        (c, m) -> Optional.of(BlockHeadersMessage.create(dummyBHList(DEFAULT_COUNT + 1))));
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByHashTaskTest.java:145]: DEFAULT_COUNT + 1 is what makes it "too many", right? 
can you move the related comment here

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/BlockPropagationManager.java (hunk 13 above)
Comment (Gabriel-Trintinalia) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/BlockPropagationManager.java:500]: maybe we could put this fix in a different PR?

**Review by macfarla — APPROVED** [review]
